src/main/java/com/tabnine/inline/InlineCompletionHandler.java

```diff
@@ -2,208 +2,104 @@
+public class InlineCompletionHandler {
+  private final CompletionFacade completionFacade;
+  private final BinaryRequestFacade binaryRequestFacade;
   private Future<?> lastPreviewTask = null;
-  private final boolean myForward;
-
-  private static boolean isLocked;
 
-  InlineCompletionHandler(boolean forward) {
-    this.myForward = forward;
+  public InlineCompletionHandler(
+      CompletionFacade completionFacade, BinaryRequestFacade binaryRequestFacade) {
+    this.completionFacade = completionFacade;
+    this.binaryRequestFacade = binaryRequestFacade;
   }
 
-  void invoke(@NotNull Editor editor, @NotNull PsiFile file, int offset) {
-    Document document = editor.getDocument();
-
-    // if we cannot modify this file, return
-    if (editor.isViewer() || document.getRangeGuard(offset, offset) != null) {
-      document.fireReadOnlyModificationAttempt();
-      EditorModificationUtil.checkModificationAllowed(editor);
-      return;
-    }
+  public void retrieveAndShowCompletion(@NotNull Editor editor, int offset) {
+    long modificationStamp = editor.getDocument().getModificationStamp();
 
-    if (isInTheMiddleOfWord(document, offset)) {
-      return;
-    }
-
-    // data about previous completions, or just a new fresh state for the editor
-    final CompletionState completionState = CompletionState.findOrCreateCompletionState(editor);
-    int lastDisplayedCompletionIndex = completionState.lastDisplayedCompletionIndex;
-
-    boolean noOldSuggestion = lastDisplayedCompletionIndex == -1;
-    boolean editorLocationHasChanged = completionState.lastStartOffset != offset;
-    boolean documentChanged =
-        completionState.lastModificationStamp != document.getModificationStamp();
-
-    if (noOldSuggestion || editorLocationHasChanged || documentChanged) {
-      // start a new query
-      completionState.lastDisplayedCompletionIndex = -1;
-      retrieveAndShowInlineCompletion(editor, file, completionState, offset);
+    if (ApplicationManager.getApplication().isUnitTestMode()) {
+      List<TabNineCompletion> completions = retrieveInlineCompletion(editor, offset);
+      rerenderCompletion(editor, completions, offset, modificationStamp);
     } else {
-      showInlineCompletion(editor, file, completionState, completionState.lastStartOffset);
-    }
-  }
+      ObjectUtils.doIfNotNull(lastPreviewTask, task -> task.cancel(false));
 
-  private boolean isInTheMiddleOfWord(@NotNull Document document, int offset) {
-    try {
-      if (DocumentUtil.isAtLineEnd(offset, document)) {
-        return false;
-      }
-      char nextChar = document.getText(new TextRange(offset, offset + 1)).charAt(0);
-      return !CLOSING_CHARACTERS.contains(nextChar) && !Character.isWhitespace(nextChar);
-    } catch (Throwable e) {
-      Logger.getInstance(getClass())
-          .debug("Could not determine if text is in the middle of word, skipping: ", e);
+      lastPreviewTask =
+          AppExecutorUtil.getAppExecutorService()
```

Comment:
please explain in a comment or something why is this extracted to a thread, while `renderCompletions` calls `invokeLater`
(i know we talked about it, but since this is not trivial it'd be nice to document it)

Reply:
TBH, I'm not sure why. I never tried doing it without it. Will do now and see how that goes.

Reply:
Tried it, it causes things to jitter, mainly but only on new lines (maybe because it takes longer?)

https://user-images.githubusercontent.com/69103343/170933324-8c39a10b-5151-44f8-8002-d47ea258a2ea.mov